### PR TITLE
iPad - centred popover presentation style

### LINKED ISSUE #1269: iPad - fix crash for popover/alerts
Crash 
- tap on attachment button
- tap in backups share option
check for other possible places

'Your application has presented a UIAlertController (<UIAlertController: 0x7f88e2872000>) of style UIAlertControllerStyleActionSheet from FlowCrypt.SideMenuNavigationController (<FlowCrypt.SideMenuNavigationController: 0x7f88e4880c00>). The modalPresentationStyle of a UIAlertController with this style is UIModalPresentationPopover. You must provide location information for this popover through the alert controller's popoverPresentationController. You must provide either a sourceView and sourceRect or a barButtonItem.  If this information is not known when you present the alert controller, you may provide it in the UIPopoverPresentationControllerDelegate method -prepareForPopoverPresentation.'

Screenshot 2022-01-04 at 19 39 46

### PR DESCRIPTION
**Pull request by Kharchevskyi**
This PR fixes crash when showing UIAlertController and UIActivityViewController

close #1269

------------------------

----------

**Tests** _(delete all except exactly one)_:
- Does not need tests (UI changes) 

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [ ] addresses the issue it closes (if any)
- [ ] code is readable and understandable
- [ ] is accompanied with tests, or tests are not needed
- [ ] is free of vulnerabilities
- [ ] is documented clearly and usefully, or doesn't need documentation

### CODE REVIEW

File: FlowCrypt/Common UI/AttachmentManager.swift [hunk 1]
```diff
@@ -17,18 +18,21 @@ final class AttachmentManager: NSObject {
...
     }
 
     private func showFileSharedAlert(with url: URL) {
+        guard let view = controller?.view else { return }
         let alert = UIAlertController(
             title: "message_attachment_saved_successfully_title".localized,
             message: "message_attachment_saved_successfully_message".localized,
             preferredStyle: .alert
-        )
+        ).popoverPresentation(style: .centred(view))
```
Comment (sosnovsky) [inline FlowCrypt/Common UI/AttachmentManager.swift:35]: There is no need for `popoverPresentation` modifier for `UIAlertController` with `preferredStyle: .alert`, it's needed only for `.actionSheet`.
Alerts are automatically centred in the parent view.

File: FlowCrypt/Controllers/Compose/ComposeViewController.swift [hunk 2]
```diff
@@ -1060,7 +1063,7 @@ extension ComposeViewController {
             title: "compose_password_modal_title".localized,
             message: "compose_password_modal_message".localized,
             preferredStyle: .alert
-        )
+        ).popoverPresentation(style: .centred(view))
```
Comment (sosnovsky) [inline FlowCrypt/Controllers/Compose/ComposeViewController.swift:1066]: `popoverPresentation` not needed here too

File: FlowCrypt/Controllers/Compose/ComposeViewController.swift [hunk 3]
```diff
@@ -1266,7 +1270,8 @@ extension ComposeViewController {
             title: "files_picking_no_library_access_error_title".localized,
             message: "files_picking_no_library_access_error_message".localized,
             preferredStyle: .alert
-        )
+        ).popoverPresentation(style: .centred(view))
```
Comment (sosnovsky) [inline FlowCrypt/Controllers/Compose/ComposeViewController.swift:1273]: not needed

File: FlowCrypt/Controllers/Compose/ComposeViewController.swift [hunk 4]
```diff
@@ -1288,7 +1293,8 @@ extension ComposeViewController {
             title: "files_picking_no_camera_access_error_title".localized,
             message: "files_picking_no_camera_access_error_message".localized,
             preferredStyle: .alert
-        )
+        ).popoverPresentation(style: .centred(view))
```
Comment (sosnovsky) [inline FlowCrypt/Controllers/Compose/ComposeViewController.swift:1296]: not needed

File: FlowCrypt/Controllers/Compose/ComposeViewController.swift [hunk 5]
```diff
@@ -1329,7 +1335,8 @@ extension ComposeViewController {
             title: "error".localized,
             message: "compose_missing_contacts_scopes".localizeWithArguments(scopes),
             preferredStyle: .alert
-        )
+        ).popoverPresentation(style: .centred(view))
```
Comment (sosnovsky) [inline FlowCrypt/Controllers/Compose/ComposeViewController.swift:1338]: not needed

File: FlowCrypt/Controllers/Setup/SetupCreatePassphraseAbstractViewController.swift [hunk 6]
```diff
@@ -152,7 +152,7 @@ extension SetupCreatePassphraseAbstractViewController {
                     title: "Pass Phrase",
                     message: "Confirm Pass Phrase",
                     preferredStyle: .alert
-                )
+                ).popoverPresentation(style: .centred(self.view))
```
Comment (sosnovsky) [inline FlowCrypt/Controllers/Setup/SetupCreatePassphraseAbstractViewController.swift:155]: not needed

File: FlowCrypt/Controllers/Threads/AlertsFactory.swift [hunk 7]
```diff
@@ -7,21 +7,24 @@
...
-        onCompletion: @escaping PassPhraseCompletion,
-        title: String = "setup_enter_pass_phrase".localized
+        onCompletion: @escaping PassPhraseCompletion
     ) -> UIAlertController {
         let alert = UIAlertController(
             title: title,
             message: nil,
             preferredStyle: .alert
-        )
+        ).popoverPresentation(style: .centred(view))
```
Comment (sosnovsky) [inline FlowCrypt/Controllers/Threads/AlertsFactory.swift:26]: not needed

File: FlowCrypt/Controllers/Threads/MessageActionsHandler.swift [hunk 8]
```diff
@@ -115,7 +115,8 @@ extension MessageActionsHandler where Self: UIViewController {
             title: "Are you sure?",
             message: "You're about to permanently delete a message",
             preferredStyle: .alert
-        )
+        ).popoverPresentation(style: .centred(view))
```
Comment (sosnovsky) [inline FlowCrypt/Controllers/Threads/MessageActionsHandler.swift:118]: not needed

File: FlowCryptCommon/Extensions/UIViewControllerExtensions.swift [hunk 9]
```diff
@@ -67,7 +72,12 @@ public extension UIViewController {
     ) {
         self.view.hideAllToasts()
         hideSpinner()
-        let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
+        let alert = UIAlertController(
+            title: title,
+            message: message,
+            preferredStyle: .alert
+        ).popoverPresentation(style: .centred(view))
```
Comment (sosnovsky) [inline FlowCryptCommon/Extensions/UIViewControllerExtensions.swift:79]: not needed

File: FlowCryptCommon/Extensions/UIViewControllerExtensions.swift [hunk 10]
```diff
@@ -53,7 +53,12 @@ public extension UIViewController {
     func showAlert(title: String? = "error".localized, message: String, onOk: (() -> Void)? = nil) {
         self.view.hideAllToasts()
         hideSpinner()
-        let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
+        let alert = UIAlertController(
+            title: title,
+            message: message,
+            preferredStyle: .alert
+        ).popoverPresentation(style: .centred(view))
```
Comment (sosnovsky) [inline FlowCryptCommon/Extensions/UIViewControllerExtensions.swift:60]: not needed

File: FlowCrypt/Functionality/Services/AppStartup.swift [hunk 11]
```diff
@@ -135,7 +135,16 @@ struct AppStartup {
...
-        let alert = UIAlertController(title: "Startup Error", message: "\(error.localizedDescription)", preferredStyle: .alert)
+        let alert = UIAlertController(
+            title: "Startup Error",
+            message: "\(error.localizedDescription)",
+            preferredStyle: .alert
+        )
+
+        if let view = window.rootViewController?.view {
+            alert.popoverPresentation(style: .centred(view))
+        }
```
Comment (sosnovsky) [inline FlowCrypt/Functionality/Services/AppStartup.swift:146]: not needed

File: FlowCrypt/Controllers/Threads/AlertsFactory.swift [hunk 12]
```diff
@@ -7,21 +7,24 @@
...
 import UIKit
+import FlowCryptCommon
 
 enum AlertsFactory {
     typealias PassPhraseCompletion = ((String) -> Void)
     typealias CancelCompletion = (() -> Void)
 
     static func makePassPhraseAlert(
+        title: String = "setup_enter_pass_phrase".localized,
+        in view: UIView,
```
Comment (sosnovsky) [inline FlowCrypt/Controllers/Threads/AlertsFactory.swift:18]: `view` parameter can be removed, as `popoverPresentation` isn't needed for `alert` here

**Review by sosnovsky — CHANGES_REQUESTED** [review]

File: FlowCrypt/Common UI/AttachmentManager.swift (hunk 1 above)
Comment (Kharchevskyi) [reply FlowCrypt/Common UI/AttachmentManager.swift:35]: @sosnovsky please check issue and attached screenshot.

File: FlowCrypt/Controllers/Compose/ComposeViewController.swift (hunk 2 above)
Comment (Kharchevskyi) [reply FlowCrypt/Controllers/Compose/ComposeViewController.swift:1066]: You will get a crash in iPad it popoverPresentationStyle is not set

File: FlowCrypt/Controllers/Compose/ComposeViewController.swift (hunk 3 above)
Comment (Kharchevskyi) [reply FlowCrypt/Controllers/Compose/ComposeViewController.swift:1273]: same

File: FlowCrypt/Controllers/Compose/ComposeViewController.swift (hunk 4 above)
Comment (Kharchevskyi) [reply FlowCrypt/Controllers/Compose/ComposeViewController.swift:1296]: same

File: FlowCrypt/Controllers/Compose/ComposeViewController.swift (hunk 5 above)
Comment (Kharchevskyi) [reply FlowCrypt/Controllers/Compose/ComposeViewController.swift:1338]: same

File: FlowCrypt/Controllers/Threads/MessageActionsHandler.swift (hunk 8 above)
Comment (Kharchevskyi) [reply FlowCrypt/Controllers/Threads/MessageActionsHandler.swift:118]: same

File: FlowCrypt/Controllers/Setup/SetupCreatePassphraseAbstractViewController.swift (hunk 6 above)
Comment (Kharchevskyi) [reply FlowCrypt/Controllers/Setup/SetupCreatePassphraseAbstractViewController.swift:155]: same

File: FlowCrypt/Controllers/Threads/AlertsFactory.swift (hunk 7 above)
Comment (Kharchevskyi) [reply FlowCrypt/Controllers/Threads/AlertsFactory.swift:26]: same

File: FlowCryptCommon/Extensions/UIViewControllerExtensions.swift (hunk 9 above)
Comment (Kharchevskyi) [reply FlowCryptCommon/Extensions/UIViewControllerExtensions.swift:79]: same

File: FlowCrypt/Functionality/Services/AppStartup.swift (hunk 11 above)
Comment (Kharchevskyi) [reply FlowCrypt/Functionality/Services/AppStartup.swift:146]: same

File: FlowCrypt/Controllers/Threads/AlertsFactory.swift (hunk 12 above)
Comment (Kharchevskyi) [reply FlowCrypt/Controllers/Threads/AlertsFactory.swift:18]: please check issue in PR description #1269
Just as an example of similar issue 
https://stackoverflow.com/questions/31577140/uialertcontroller-is-crashed-ipad

Comment (sosnovsky) [reply FlowCrypt/Controllers/Threads/AlertsFactory.swift:18]: Yes, I read original issue, error there says `Your application has presented a UIAlertController (<UIAlertController: 0x7f88e2872000>) of style UIAlertControllerStyleActionSheet`

The same for stackoverflow post, it shows `.actionSheet`, not `.alert`:
`let shareMenu = UIAlertController(title: nil, message: "Share using", preferredStyle: .ActionSheet)`

So we need `popoverPresentationStyle` set only for UIAlertController with `preferredStyle: .actionSheet`

I just tested it on iPad - `preferredStyle: .alert` works well without `popoverPresentationStyle`

Comment (Kharchevskyi) [reply FlowCrypt/Controllers/Threads/AlertsFactory.swift:18]: strange why I have this error on .alert

Comment (Kharchevskyi) [reply FlowCrypt/Controllers/Threads/AlertsFactory.swift:18]: are you testing it on simulator?

Comment (sosnovsky) [reply FlowCrypt/Controllers/Threads/AlertsFactory.swift:18]: Yes, on simulator

Comment (Kharchevskyi) [reply FlowCrypt/Controllers/Threads/AlertsFactory.swift]: Looks like indeed not needed for alert style. Please check latest changes

**Review by sosnovsky — APPROVED** [review]
Looks good now 👍

**Pull request merged**